integration/non_root_user_cap_test.go

```diff
@@ -0,0 +1,84 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+func TestNonRootUserCap(t *testing.T) {
+	for name, test := range map[string]struct {
+		uid     int64
+		startOK bool
+	}{
+		"shouldn't be able to run test container, with a private workdir that is accessible to nobody, as non-root user": {
+			uid:     1234,
+			startOK: false,
+		},
+		"should be able to run test container, with a private workdir that is accessible to nobody, as root user": {
+			uid:     0,
+			startOK: true,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			t.Log("Create a sandbox")
+			sbConfig := PodSandboxConfig("sandbox", "non-root-user-cap")
+			sb, err := runtimeService.RunPodSandbox(sbConfig, *runtimeHandler)
+			require.NoError(t, err)
+			// Make sure the sandbox is cleaned up.
+			defer func() {
+				assert.NoError(t, runtimeService.StopPodSandbox(sb))
+				assert.NoError(t, runtimeService.RemovePodSandbox(sb))
+			}()
+
+			const (
+				testImage     = "gcr.io/k8s-cri-containerd/private-workdir:1.0"
+				containerName = "test-container"
+			)
+			t.Logf("Pull test image %q", testImage)
+			img, err := imageService.PullImage(&runtime.ImageSpec{Image: testImage}, nil, sbConfig)
+			require.NoError(t, err)
+			defer func() {
+				assert.NoError(t, imageService.RemoveImage(&runtime.ImageSpec{Image: img}))
+			}()
+
+			t.Log("Create a container to test capabilities")
+			cnConfig := ContainerConfig(
+				containerName,
+				testImage,
+				WithCommand("ls", "/"),
```

Comment:
I think this should be `ls` (or `ls .`), rather than `ls /`, since you want to check the /workdir (which is cwd), not the root directory.

This is also the reason why `ls /` doesn't work as expected, but `touch ./foo` does.

Reply:
Ack, agree thanks.
Sorry I fell off on this, went OOO and got distracted with other things. I'll get back on this later today.

Reply:
Ah so actually in a later commit I moved this to `touch ./foo` instead, I think the `"ls", "/"` came from the original CRI PR I picked up.

Reply:
I switched it to `ls .` anyhow, I believe the issue persists where currently `crun` will succeed but not `runc` without the last commit (checking for container run to success instead of just container start success) due to the ordering of chroot vs drop cap.